File: bench/Reference/binary_read_chunks.java
import java.io.*;

class binary_read_chunks {
    public static void main(String[] args) throws Exception {
        FileInputStream fis = new FileInputStream("bench/data/bench_binary.bin");
        byte[] buf = new byte[65536];
        long total = 0;
        int n;
        while ((n = fis.read(buf)) > 0)
            total += n;
        fis.close();
        System.out.println(total);
    }
}

File: bench/Reference/binary_read_chunks.hs
-- Haskell reference: binary chunked read
import System.IO
import qualified Data.ByteString as BS

main :: IO ()
main = do
    h <- openBinaryFile "bench/data/bench_binary.bin" ReadMode
    total <- readChunks h 0
    hClose h
    print total
  where
    readChunks h acc = do
        chunk <- BS.hGet h 65536
        if BS.null chunk
            then return acc
            else readChunks h (acc + BS.length chunk)
